File: handler/node/notitg.js
const { NotITG } = require('notitg-external')
const { EventEmitter } = require('events')

function chunk(arr, n = 26) {
	let a = []
	for (let i = 0; i < arr.length; i += n) {
		a.push(arr.slice(i, i + n))
	}
	return a
}

class NotITGWrapper extends NotITG {
	/**
	 * @type {NodeJS.Timer}
	 */
	__running = undefined

	constructor(appID) {
		super()
		this.appID = appID
		this.events = new EventEmitter()
	}

	/**
	 * @type {number}
	 */
	__pid = undefined
	/**
	 * @type {boolean}
	 */
	__deep = undefined
	/**
	 * @param {?boolean} deep Scan all programs
	 * @param {?number} pid NotITG's Process ID
	 */
	Start(deep, pid) {
		if (this.__running) return
		if (pid) this.__pid = pid
		if (deep) this.__deep = deep

		this.__running = setInterval(() => this.Tick(), 10)
	}

	Stop() {
		if (!this.__running) return
		this.Disconnect()
		clearInterval(this.__running)
	}

	// -- //
	/**
	 * @type {{buffer: number[], set: number}[]}
	 */
	__writeBuffer = []
	/**
	 * @type {number[]}
	 */
	__readBuffer = []
	/**
	 * Send data to NotITG
	 * @param {number[]} buffer The buffer to send
	 */
	Write(buffer) {
		if (buffer.length <= 26) {
			this.__writeBuffer.push({
				buffer: [...buffer],
				set: 0,
			})
		} else {
			const chunks = chunk(buffer)
			for (let i = 0; i < chunks.length; i++) {
				console.log(chunks[i].join(' '))
				this.__writeBuffer.push({
					buffer: chunks[i],
					set: i + 1 === chunks.length ? 2 : 1,
				})
			}
		}
	}

	__heartbeatStatus = 0
	initialized = false
	Tick() {
		if (!this.Heartbeat()) {
			if (
				this.__heartbeatStatus !== 2 &&
				((this.__pid && this.FromProcessID(this.__pid)) || (!this.__pid && this.Scan(this.__deep)))
			) {
				if (['V1', 'V2'].includes(this.Version)) {
					console.log(`⚠ Unsupported NotITG version! Expected V3+, got ${this.Version}.`)
					this.Stop()
				}

				this.__heartbeatStatus = 2
				this.initialized = false

				/*console.log('> --------------------------------')
				console.log('✅ Found NotITG!')
				console.log(`Version >> ${this.Version}`)
				console.log('> --------------------------------')*/
				this.events.emit('connect')
			} else if (this.__heartbeatStatus == 0) {
				this.__heartbeatStatus = 1

				//console.log('❌ Could not find a version of NotITG!')
			} else if (this.__heartbeatStatus === 2) {
				this.__heartbeatStatus = 0

				this.Disconnect()
				//console.log('❓ NotITG has exited')
				this.events.emit('disconnect')
			}
		} else {
			if (this.GetExternal(60) === 0) return
			if (!this.initialized) {
				// console.log('🏴 NotITG has initialized!')
				this.events.emit('initialized')
				this.initialized = true
			}

			// READ
			if (this.GetExternal(57) === 1 && this.GetExternal(59) === this.appID) {
				/**
				 * @type {number[]}
				 */
				const buffer = []

				for (let i = 28; i < 28 + this.GetExternal(54); i++) {
					buffer.push(this.GetExternal(i))
					this.SetExternal(i, 0)
				}

				const stat = this.GetExternal(55)
				this.events.emit('read-buffer', buffer, stat)
				if (stat === 0) this.events.emit('read', buffer)
				else {
					this.__readBuffer.push(...buffer)
					if (stat === 2) {
						this.events.emit('read', this.__readBuffer)
						this.__readBuffer = []
					}
				}

				this.SetExternal(54, 0)
				this.SetExternal(55, 0)
				this.SetExternal(59, 0)
				this.SetExternal(57, 0)
			}

			// WRITE
			if (this.__writeBuffer.length > 0 && this.GetExternal(56) === 0) {
				this.SetExternal(56, 1)

				const { buffer, set } = this.__writeBuffer.shift()

				for (let i = 0; i < buffer.length; i++) {
					this.SetExternal(i, buffer[i])
				}
				this.SetExternal(26, buffer.length)
				this.SetExternal(27, set)
				this.SetExternal(56, 2)
				this.SetExternal(58, this.appID)

				this.events.emit('write', buffer, set)
			}
		}
	}
}

const ENCODE_GUIDE = `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 \n'\"~!@#$%^&*()<>/-=_+[]:;.,\`{}`
/**
 * Encodes the string into a buffer
 * @param {string} str The string to be encoded
 * @returns {number[]}
 */
function Encode(str) {
	return str.split('').map((char) => ENCODE_GUIDE.indexOf(char) + 1)
}
/**
 * Decodes the buffer into a string
 * @param {number[]} buffer
 * @returns {string}
 */
function Decode(buffer) {
	return buffer.map((buff) => ENCODE_GUIDE[buff - 1]).join('')
}

module.exports = { NotITGWrapper, Encode, Decode }

File: handler/dotnet/NotITG.cs
using System;
using System.Collections.Generic;
using System.Timers;
using System.Linq;

namespace NotITG
{
	public partial class NotITGWrapper
	{
		public int applicationID;
		public NotITGWrapper(int appID)
		{
			applicationID = appID;

			t = new Timer();
			t.Elapsed += (e, o) => { Tick(); };
			t.Interval = 10;
			t.Enabled = false;

			notitgClient = new External.NotITG();
		}

		public class BufferEventArgs : EventArgs
		{
			public int[] Buffer { get; set; }
		}
		public delegate void BufferEventHandler(object sender, BufferEventArgs e);

		public event BufferEventHandler OnRead;
		public event BufferEventHandler OnWrite;
		public event BufferEventHandler OnBufferRead;

		public event EventHandler OnConnect;
		public event EventHandler OnInitialized;
		public event EventHandler OnDisconnect;

		private bool scanDeep = false;
		private int? scanProcessID;
		public void Start(bool deep = false, int? pid = null)
		{
			if (t.Enabled) return;

			scanDeep = deep;
			scanProcessID = pid;
			t.Start();
		}

		private Timer t;
		private External.NotITG notitgClient;
		public bool IsConnected
		{
			get
			{
				if (notitgClient == null) return false;
				return notitgClient.Connected;
			}
		}

		public string GamePath
		{
			get
			{
				if (notitgClient == null) return "";
				return notitgClient.GamePath;
			}
		}

		public void Stop()
		{
			if (!t.Enabled) return;
			notitgClient.Disconnect();
			t.Stop();
		}

		private List<WriteStruct> WriteBuffer = new List<WriteStruct>();
		private List<int> ReadBuffer = new List<int>();
		private struct WriteStruct
		{
			public enum SetStatus
			{
				SINGLE = 0,
				SET = 1,
				SET_END = 2,
			}
			public List<int> buffer;
			public SetStatus set;
		}

		public void Write(int[] buffer)
		{
			if (!notitgClient.Connected) return;
			if( buffer.Length <= 26 )
			{
				WriteBuffer.Add(new WriteStruct()
				{
					buffer = buffer.ToList(),
					set = WriteStruct.SetStatus.SINGLE,
				});
			}
			else
			{
				for(int i = 0; i < buffer.Length; i += 26)
				{
					WriteBuffer.Add(new WriteStruct()
					{
						buffer = buffer.ToList().GetRange(i, Math.Max(26, buffer.Length - 1 - i)),
						set = i + 26 >= buffer.Length ? WriteStruct.SetStatus.SET_END : WriteStruct.SetStatus.SET
					});
				}
			}
		}

		private string ENCODE_GUIDE = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 \n'\"~!@#$%^&*()<>/-=_+[]:;.,`{}";
		public int[] Encode(string str)
		{
			return str.ToCharArray().Select(ch => ENCODE_GUIDE.IndexOf(ch) + 1).ToArray();
		}
		public string Decode(int[] str)
		{
			return string.Join("", str.Select(index => ENCODE_GUIDE[index - 1]).ToArray());
		}

		private enum HEART
		{
			DISCONNECTED = 0,
			SEARCHING = 1,
			CONNECTED = 2,
		}
		private HEART heartbeatStatus = HEART.DISCONNECTED;
		private bool initialized = false;
		private void Tick()
		{
			// Temporary disable the timer
			// .Heartbeat() doesn't "temporarily" stop the timer "thread" (i dont know if this is the right term).
			// The Timer WILL run the Tick function regardless of whether it has finished the previous tick.
			t.Enabled = false;

			if (!notitgClient.Heartbeat())
			{
				if(
					heartbeatStatus != HEART.CONNECTED &&
					(
						(scanProcessID != null && notitgClient.ScanFromProcessID((int)scanProcessID)) ||
						(scanProcessID == null && notitgClient.Scan(scanDeep))
					)
				)
				{
					if(notitgClient.Version == External.Details.NotITGVersionNumber.V1 || notitgClient.Version == External.Details.NotITGVersionNumber.V2)
					{
						Console.WriteLine("Unsupported NotITG Version! Expected at least V3");
						Stop();
						return;
					}
					heartbeatStatus = HEART.CONNECTED;
					initialized = false;

					OnConnect?.Invoke(this, null);
				}
				else if(heartbeatStatus == HEART.DISCONNECTED)
				{
					heartbeatStatus = HEART.SEARCHING;
				}
				else if(heartbeatStatus == HEART.CONNECTED)
				{
					notitgClient.Disconnect();
					heartbeatStatus = HEART.DISCONNECTED;
					OnDisconnect?.Invoke(this, null);
				}
			}
			else if(notitgClient.GetExternal(60) != 0)
			{
				if(!initialized)
				{
					OnInitialized?.Invoke(this, null);
					initialized = true;
				}

				// READ
				if( notitgClient.GetExternal(57) == 1 && notitgClient.GetExternal(59) == applicationID)
				{
					int length = notitgClient.GetExternal(54);
					int[] buffer = new int[length];

					for( int i = 28; i < 28 + length; i++ )
					{
						buffer[i - 28] = notitgClient.GetExternal(i);
						notitgClient.SetExternal(i, 0);
					}

					int stat = notitgClient.GetExternal(55);
					OnBufferRead?.Invoke(this, new BufferEventArgs() { Buffer = buffer });
					if( stat == 0 )
					{
						OnRead?.Invoke(this, new BufferEventArgs() { Buffer = buffer });
					}
					else
					{
						ReadBuffer.AddRange(buffer);
						if( stat == 2 )
						{
							OnRead?.Invoke(this, new BufferEventArgs() { Buffer = ReadBuffer.ToArray() });
							ReadBuffer.Clear();
						}
					}
				}

				// WRITE
				if( WriteBuffer.Count > 0 && notitgClient.GetExternal(56) == 0 )
				{
					notitgClient.SetExternal(56, 1);

					var WriteContent = WriteBuffer[0];
					WriteBuffer.RemoveAt(0);

					for( int i = 0; i < WriteContent.buffer.Count; i++ )
					{
						notitgClient.SetExternal(i, WriteContent.buffer[i]);
					}

					notitgClient.SetExternal(26, WriteContent.buffer.Count);
					notitgClient.SetExternal(27, (int)WriteContent.set);
					notitgClient.SetExternal(56, 2);
					notitgClient.SetExternal(58, applicationID);

					OnWrite?.Invoke(this, new BufferEventArgs()
					{
						Buffer = WriteContent.buffer.ToArray()
					});
				}
			}

			t.Enabled = true;
		}
	}
}
